gramineproject/gramine
# Rewrite sysfs topology support

## Description of the changes

This PR completely rewrites all three parts of sysfs topology handling:
- parsing sysfs files on the host (`topo_info.c`),
- passing them to LibOS, possibly across a security boundary (`Linux-SGX/db_main.c`),
- generating sysfs files inside LibOS based on the topology information (`fs/sys/*`).

The overall idea of the rewrite is to simplify and normalize the data structures passing through the untrusted interface, to ensure they are properly handled and sanitized. Effectively, this PR shifts as much complexity as possible to either untrusted side or to the part working on already sanitized, in-enclave data.

The reason for this change is the number of security issues found in the old code and review difficulty of sanitization in the old approach (despite multiple review passes I kept finding new issues).

I plan some more changes for future PRs, but they aren't strictly related to the code changed here (e.g. adding a proper test suite for sysfs topology).

Fixes #483.
Fixes https://github.com/gramineproject/graphene/issues/2105.

## Functional changes

- Added support for offline CPU threads (but no hot-plug, the state is snapshotted at the start).
- `/sys/devices/system/cpu/cpu<n>/topology/core_id` values have changed - they aren't socket-local anymore, but are globally unique. This shouldn't matter, as this node is documented in Linux as "the hardware platform’s identifier (rather than the kernel’s). The actual value is architecture and platform dependent.".

## Rough edges

The worst part of this PR is probably `Linux-common/topo_info.c`, but I don't see any other way to write this. This is something where we just need lambdas, but C doesn't have them. Anyways, if you have some ideas how to make it better, then please add them to the review.

Also, please be careful when reading the code around topology structures. `pal_cpu_thread_info` and `pal_numa_node_info` fields are valid _only_ for online nodes. Technically, we could remove offline nodes and pretend they don't exist (as we don't support hot-plug and probably never will), but that won't make the implementation any simpler (because we'd need to translate indices in parser code).

## Useful links for reviewers

- https://www.kernel.org/doc/html/latest/admin-guide/cputopology.html
- https://www.kernel.org/doc/html/latest/admin-guide/abi-stable.html

## sysfs mysteries

The format used in Linux sysfs is quite obscure and the documentation is partial at best. Things I don't know:
- What's the different between "die" and "package" in their terminology?
- Can a NUMA node actually be turned offline? I never succeeded in shutting down one, even after shutting down all CPUs inside the node it shows up as online on `node/online` list. If that's the case, then what's the purpose of `node/online` on Linux?

## How to test this PR?

This PR requires extensive testing, although it should be quite safe to merge - it contains less assumptions than the old implementation, at the same time refusing all internally inconsistent topology data. Despite this, there may be something I didn't think of :)

To anyone reading this who has some unusual environment (i.e. something more than boring 8-threaded single CPU) and have a moment: please try out this branch, just starting Gramine built from it on any binary should be enough. If you have more time, checking if contents of `/sys/devices/system/` make sense would be helpful.

I'm especially looking for:
- machines with more than 1 NUMA node,
- virtualized environments,
- machines with non-default configuration (e.g. some CPUs shut down or disabled).


---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/gramineproject/gramine/530)
